### PR TITLE
Get own IP when there are no peers

### PR DESCRIPTION
**Pull request by clairew**
Fixes [#4311](https://github.com/tailscale/tailscale/issues/4311)

When getting status without peers, the tailscale ips key should still be populated with the tsnet server's self ip

### CODE REVIEW

File: ipn/ipnlocal/state_test.go [hunk 1]
```diff
@@ -291,6 +292,47 @@ func (cc *mockControl) UpdateEndpoints(endpoints []tailcfg.Endpoint) {
 // network delays) are just ignored for now, which makes the test
 // predictable, but maybe a bit less thorough. This is more of an overall
 // state machine test than a test of the wgengine+magicsock integration.
+func TestStatusWithoutPeers(t *testing.T) {
```
Comment (clairew) [inline ipn/ipnlocal/state_test.go:295]: I added the test here to make use of the mock control server helper methods - but we may want to refactor and put these methods in a helper library.

Comment (maisem) [reply ipn/ipnlocal/state_test.go:295]: You don't need to be in the same file for the helpers to be accessible, so you can put the test in local_test.go and still use these.

File: ipn/ipnlocal/local.go [hunk 2]
```diff
@@ -660,16 +661,10 @@ func (b *LocalBackend) populatePeerStatusLocked(sb *ipnstate.StatusBuilder) {
...
 		}
-		var tailscaleIPs = make([]netip.Addr, 0, len(p.Addresses))
-		for _, addr := range p.Addresses {
-			if addr.IsSingleIP() && tsaddr.IsTailscaleIP(addr.Addr()) {
-				tailscaleIPs = append(tailscaleIPs, addr.Addr())
-			}
-		}
 		ps := &ipnstate.PeerStatus{
 			InNetworkMap: true,
+			TailscaleIPs: nil,
```
Comment (maisem) [inline ipn/ipnlocal/local.go:666]: remove the line

File: ipn/ipnlocal/state_test.go [hunk 3]
```diff
@@ -291,6 +292,47 @@ func (cc *mockControl) UpdateEndpoints(endpoints []tailcfg.Endpoint) {
...
+	b.Login(nil)
+	cc.send(nil, "", false, &netmap.NetworkMap{
+		MachineStatus: tailcfg.MachineAuthorized,
+		Addresses:     ipps("100.101.101.101"),
+		SelfNode: &tailcfg.Node{
+			Addresses: ipps("100.101.101.101"),
+		},
+	})
+	{
+	}
```
Comment (maisem) [inline ipn/ipnlocal/state_test.go:326]: remove

**Review by maisem — COMMENTED** [review]
We use `<project>: <description>` as the commit title to provide more info on the packages being modified.

Something like
`ipn/ipnlocal: fill out self.Addresses in StatusWithoutPeers

**Review by bradfitz — COMMENTED** [review]
I'm working on trying to understand this and how the bug came to be.

**Comment by bradfitz** [comment]
Okay, I think I understand now.

The idea with the UpdateStatus stuff is that there are a bunch of subsystems that can all participate in building it out: LocalBackend, the wgengine.Engine (userspace engine), magicsock.Conn. But with the localapi StatusWithoutPeers path, the magicsock.Conn never gets a chance to make its edit. That's the real bug. I don't think we should try to make LocalBackend paper over that by fixing it up.

Rather than changing the Engine interface's UpdateStatus method, how about we add a new method like `WantPeers() bool` on `StatusBuilder`, and then `magicsock.Conn` can check that before doing its `c.peerMap.forEachEndpoint` call.

Then LocalBackend's Status-vs-StatusWithoutPeers can be made more consistent to go down similar paths. The only difference would be when they make their StatusBuilder whether they set that bool.

**Pull request closed without merging**